### PR TITLE
closes: #355

### PR DESCRIPTION
**Pull request by DarthUdp**
fix: fixed issue where null characters in the middle of a string would cut the string short on echo.

### CODE REVIEW

**Comment by DarthUdp** [comment]
For some reason the tests fail but trying some of the tests by hand gives the expected output

**Comment by CoffeeTableEspresso** [comment]
Have you tried running all tests locally? `tests.sh` is the script for this. There's also a `-m` option to toggle whether you want to run them with `valgrind` enabled.

I suspect you've got a memory problem somewhere, but hard to tell just by looking at your code.

**Comment by DarthUdp** [comment]
Can I assume built-ins `v->on_heap == false` safely?

**Comment by DarthUdp** [comment]
locally all tests passed the remaining errors are about msvc and c++ complaining about linking symbols

**Comment by CoffeeTableEspresso** [comment]
That's not a safe assumption, `on_heap` tells where the memory for the string was allocated (depending on creation, e.g. from the C API, dynamically at runtime, from a YASL string literal, etc). it shouldn't be relied on for this sort of stuff, even tho in this case it seems to be fine.

Also, your CL is modifying the string in place, which it should not do (strings should be immutable in YASL).

Consider the following:

```
const x = 'abc\0def'

echo len x

echo x

echo len x

```

This should print `7` for the length of `x` both times, but the second time it prints `6` under your CL, because the string has been modified. (Slicing the string also shows a similar problem, where the substring is from the indices based on what `x` was originally.)

**Comment by DarthUdp** [comment]
> That's not a safe assumption, `on_heap` tells where the memory for the string was allocated (depending on creation, e.g. from the C API, dynamically at runtime, from a YASL string literal, etc). it shouldn't be relied on for this sort of stuff, even tho in this case it seems to be fine.
> 
> Also, your CL is modifying the string in place, which it should not do (strings should be immutable in YASL).
> 
> Consider the following:
> 
> ```
> const x = 'abc\0def'
> 
> echo len x
> 
> echo x
> 
> echo len x
> ```
> 
> This should print `7` for the length of `x` both times, but the second time it prints `6` under your CL, because the string has been modified. (Slicing the string also shows a similar problem, where the substring is from the indices based on what `x` was originally.)

I have adjusted the code to allocate a buffer to keep the old string immutable, ensuring it passes the tests now

**Comment by CoffeeTableEspresso** [comment]
> a bit of a doubt is it ok if I check malloc result and if it could not allocate throw like this?
> 
> ```c
> if (dest == NULL) {
> 		vm_throw_err(vm, YASL_ERROR);
> }
> ```
> 
> other uses of vm_throw_err in VM.c suggests to me that `return` is not needed or even correct, right?

I've so far coded YASL under the assumption that `malloc` will never fail, so there's not really a need to check the result. I will likely change this in the future but for now no need to bother checking.

File: interpreter/VM.c [hunk 1]
```diff
@@ -1190,7 +1190,11 @@ void vm_close_all(struct VM *const vm) {
 static void vm_ECHO(struct VM *const vm) {
 	vm_stringify_top(vm);
 	struct YASL_String *v = vm_popstr(vm);
-	vm_print_out(vm, "%.*s\n", (int)YASL_String_len(v), v->start + v->str);
+	size_t strlen = (int)YASL_String_len(v);
+	char *dest = (char *) malloc(strlen);
+	size_t copied = io_str_strip_char(dest, v->str, strlen, 0);
```
Comment (CoffeeTableEspresso) [inline interpreter/VM.c:1195]: This should be `v->str + v->start`, `v->str` points to the beginning of the underlying buffer, not the start of the string.

You can see that if you do a slice for example:
```
const x = '\0abc\0def'

echo len x

echo x
echo x[1:]

echo len x

```

**Comment by DarthUdp** [comment]
> > a bit of a doubt is it ok if I check malloc result and if it could not allocate throw like this?
> > ```c
> > if (dest == NULL) {
> > 		vm_throw_err(vm, YASL_ERROR);
> > }
> > ```
> > 
> > 
> > other uses of vm_throw_err in VM.c suggests to me that `return` is not needed or even correct, right?
> 
> I've so far coded YASL under the assumption that `malloc` will never fail, so there's not really a need to check the result. I will likely change this in the future but for now no need to bother checking.

I noticed it after I posted the comment lol, sorry for that

**Review by CoffeeTableEspresso — CHANGES_REQUESTED** [review]
Besides that one small thing, this LGTM.

File: interpreter/VM.c (hunk 1 above)
Comment (DarthUdp) [reply interpreter/VM.c:1195]: k, changing it iab

Comment (DarthUdp) [reply interpreter/VM.c]: Done ;)

**Review by CoffeeTableEspresso — APPROVED** [review]
LGTM

**Comment by CoffeeTableEspresso** [comment]
Thanks so much for your help here, contributions are always welcome in the future as well!

**Pull request merged**